metaflow/plugins/secrets/secrets_decorator.py

```diff
@@ -0,0 +1,228 @@
+import json
+import os
+import re
+from collections import defaultdict
+
+from metaflow.exception import MetaflowException
+from metaflow.decorators import StepDecorator
+from metaflow.plugins.secrets.aws_secrets_manager_secrets_provider import (
+    AwsSecretsManagerSecretsProvider,
+)
+from metaflow.plugins.secrets.inline_secrets_provider import InlineSecretsProvider
+
+SECRETS_BACKEND_PROVIDERS = {
+    "aws-secrets-manager": AwsSecretsManagerSecretsProvider(),
+    "_inline": InlineSecretsProvider(),
+}
+
+
+def get_default_secrets_backend_type():
+    from metaflow.metaflow_config import DEFAULT_SECRETS_BACKEND_TYPE
+
+    if DEFAULT_SECRETS_BACKEND_TYPE is None:
+        raise MetaflowException(
+            "No default secrets backend type configured, but needed by @secrets.  Set METAFLOW_DEFAULT_SECRETS_BACKEND_TYPE."
+        )
+    return DEFAULT_SECRETS_BACKEND_TYPE
+
+
+class SecretSpec:
+    def __init__(self, secrets_backend_type, secret_id, options={}):
+        self._secrets_backend_type = secrets_backend_type
+        self._secret_id = secret_id
+        self._options = options
+
+    @property
+    def secrets_backend_type(self):
+        return self._secrets_backend_type
+
+    @property
+    def secret_id(self):
+        return self._secret_id
+
+    @property
+    def options(self):
+        return self._options
+
+    def to_json(self):
+        """Mainly used for testing... not the same as the input dict in secret_spec_from_dict()!"""
+        return {
+            "secrets_backend_type": self.secrets_backend_type,
+            "secret_id": self.secret_id,
+            "options": self.options,
+        }
+
+    @staticmethod
+    def secret_spec_from_str(secret_spec_str):
+        # "." may be used in secret_id one day (provider specific). HOWEVER, it provides the best UX for
```

Comment:
cc @savingoyal per out convo... I think `.` is the right way to go after all.  Some jotted down the reasoning as a long comment for posterity.